### ISSUE TITLE
Kubernetes' native environment substitution doesn't work all the time

### ISSUE THREAD

**Issue by yuvipanda** [jupyterhub/kubespawner#491]
### Bug description

Kubernetes supports [environment variable substitution](https://kubernetes.io/docs/tasks/inject-data-application/define-interdependent-environment-variables/) - you can derive one variable by string interpolation of others. However, you can only interpolate variables you have already declared in the *list* of environment variables. And since kubespawner takes a dict rather than a list, we can't control the order here - it's always alphabetically sorted (not sure if that is intentional).

### To reproduce

Here's some sample YAML config for z2jh

```yaml
singleuser:
  env:
    MESSAGE_FIRST: "Hello"
    MESSAGE_SECOND: "World"
    MESSAGE: "$(MESSAGE_FIRST) $(MESSAGE_SECOND)
```

In the user container, you would expect `echo $MESSAGE` to give you `Hello World`, but it'll actually give you `$(MESSAGE_FIRST) $(MESSAGE_SECOND)` - no expansion. This is because, `MESSAGE` sorts *above* `MESSAGE_FIRST` and `MESSAGE_SECOND`, so when Kubernetes tries to process the expansions, it fails since these aren't defined yet.

If you instead change it to this, to force ordering:
```yaml
singleuser:
  env:
    MESSAGE_FIRST: "Hello"
    MESSAGE_SECOND: "World"
    ZESSAGE: "$(MESSAGE_FIRST) $(MESSAGE_SECOND)
```

The `ZESSAGE` is set to `Hello World` correctly.

**Comment by consideRatio** [comment]
We could introduce an algorithm to stop using alphabetical order, but instead try to ensure we avoid this issue?

1. Expand all values using KubeSpawner expansions
2. Put anything not matching `"$(...)"
3. Loop until the remaining envs are 0:
   1. If all remaining envs reference other remaining envs, a circular dependency can be assumed and we emit a warning and put the envs in any order.
   2. Else put the first env remaining not referencing other remaining envs

**Comment by yuvipanda** [comment]
Hmm, this requires we keep track of what features kubernetes supports and keep it up to date. Instead, I'd prefer some way of either letting environment be a list instead of string, or a 'weight's key that can be used. All keys added by KubeSpawner will be weight 0, z2jh weight 1, and then every other key is user specified weight. This lets users specify things as they wish and control ordering for whatever reason they might want. This already happens elsewhere (`hub.extraConfig` for example), so would probably work ok for most cases?

**Comment by consideRatio** [comment]
So you suggest making `KubeSpawner.environment` be allowed to be a list and a dict, overriding the default implementation in `Spawner.environment`?

I think this may cause a headache both in the added complexity in KubeSpawner, z2jh, and for users.

I lean towards preferring a smarter sorting by default instead of an alphabetical as a more attractive compromise.

I also think it would be reasonable to document the following approach which I think would work and now realize. So in #426 we started supporting k8s native env var specification, such as `{name: MY_ENV, value: some-value}` objects. And, that following #426, inclusion of an explicit `name` field would make the top level key be irrelevant - but still probably what we sort alphabetically using.

In other words, I think this would work

```yaml
singleuser:
  extraEnv:
    00-my-env:
      name: ACTUAL_ENV_NAME_Z
      value: blablabla
    zz-my-other-env:
      name: ACTUAL_ENV_NAME_A
      value: blablabla
```

### Related

[Z2JH docs](https://zero-to-jupyterhub.readthedocs.io/en/latest/resources/reference.html#singleuser-extraenv)

https://github.com/jupyterhub/kubespawner/blob/386a24c8085d90de21c498c50484db4a236479c9/kubespawner/objects.py#L438-L449

**Comment by yuvipanda** [comment]
aaaah, I didn't realize you can specify name directly. That worked very well - see https://github.com/2i2c-org/pilot-hubs/pull/285. Thanks a lot for thinking this through, @consideratio!
